### PR TITLE
Move [`assertions_on_result_states`] to restriction

### LINKED ISSUE #9263: `assertions_on_result_states` leads to less clear tests
### Summary

IMO using `assert!(foo().is_err())` is clearer within a test to imply that the assertion is the point of the test rather than `foo().unwrap_err()`.

### Lint Name

assertions_on_result_states

### Reproducer

I tried [this code](https://github.com/Nullus157/stylish-rs/blob/e92b0bc2e223f9d6c40affc31e9b6744b327dd92/tests/tests.rs#L58-L71):

```rust
    #[test]
    fn error() {
        struct Foo;

        impl stylish::Display for Foo {
            fn fmt(&self, _: &mut stylish::Formatter<'_>) -> std::fmt::Result {
                Err(std::fmt::Error)
            }
        }

        let mut s = Vec::<u8>::new();
        let mut writer = stylish::io::Plain::new(&mut s);
        assert!(stylish::writeln!(writer, "{:s}", Foo).is_err());
    }
```

I [saw this happen](https://github.com/Nullus157/stylish-rs/runs/7588631386?check_suite_focus=true):

```
error: called `assert!` with `Result::is_err`
Error:   --> tests/tests.rs:70:9
   |
70 |         assert!(stylish::writeln!(writer, "{:s}", Foo).is_err());
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `stylish::writeln!(writer, "{:s}", Foo).unwrap_err()`
   |
   = note: `-D clippy::assertions-on-result-states` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#assertions_on_result_states
```

I expected to see this happen:

### Version

```text
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: latest update on 2022-07-30, rust version 1.64.0-nightly (3924dac7b 2022-07-29)
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by tabokie**
Close #9263

This lint causes regression on readability of code and log output. And printing runtime values is not particularly useful for majority of tests which should be reproducible.

changelog: Move [`assertions_on_result_states`] to restriction and don't lint it for unit type

Signed-off-by: tabokie <xy.tao@outlook.com>

### CODE REVIEW

**Comment by xFrednet** [comment]
I agree with these changes @flip1995, the lint was added 10 days ago in #9225. If we merge it now, it won't make it on the beta branch, if I'm correct. Do we maybe want to do the change in rust-lang/rust or backport it later?

(This is another place where nightly lints would be so helpful. It's on my todo list after `lint_reasons` has been stabilized :sweat_smile: )

**Comment by flip1995** [comment]
We would need an out-of-cycle sync to get it into the rust repo before the release. I'm good with that, as long as it doesn't add too many other things (like new lints). 

We can also just do the `restriction` move in the Rust repo and let the other changes in this PR ride the train. That way, we don't need a sync.

File: clippy_lints/src/assertions_on_result_states.rs [hunk 1]
```diff
@@ -96,3 +100,15 @@ fn has_debug_impl<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
         .get_diagnostic_item(sym::Debug)
         .map_or(false, |debug| implements_trait(cx, ty, debug, &[]))
 }
+
+fn is_unit_type(ty: Ty<'_>) -> bool {
```
Comment (flip1995) [inline clippy_lints/src/assertions_on_result_states.rs:104]: Can't you call `ty.is_unit() || ty.is_never()` instead of this function?

**Review by flip1995 — COMMENTED** [review]
Only one comment, otherwise LGTM

File: clippy_lints/src/assertions_on_result_states.rs (hunk 1 above)
Comment (tabokie) [reply clippy_lints/src/assertions_on_result_states.rs:104]: Good idea. Was copying it from another older lint before.

Comment (flip1995) [reply clippy_lints/src/assertions_on_result_states.rs]: Oh which one? We should also fix it there... (not in this PR though.

**Comment by flip1995** [comment]
@bors r+

Thanks!

File: clippy_lints/src/assertions_on_result_states.rs (hunk 1 above)
Comment (tabokie) [reply clippy_lints/src/assertions_on_result_states.rs]: Here: https://github.com/rust-lang/rust-clippy/blob/7c427f0fdadf07c117d23d0b3352a4870c9b17d3/clippy_lints/src/map_unit_fn.rs#L99

Comment (flip1995) [reply clippy_lints/src/assertions_on_result_states.rs]: Are you motivated to fix it also there and open a PR for it? :)

Comment (tabokie) [reply clippy_lints/src/assertions_on_result_states.rs]: Sure, will fix it tomorrow.

**Pull request merged**